common/memory-stream/memory-stream.cpp

```diff
@@ -198,6 +198,81 @@ class FaultStream : public IStream
 
 };
 
+using namespace photon::net;
+class StringSocketStreamImpl : public StringSocketStream {
+public:
+    virtual ssize_t recv(void *buf, size_t count, int flags = 0) override {
+        size_t n = rand() % 8 + 1;
+        if (n > _inv.size()) n = _inv.size();
+        if (n > count) n = count;
+        memcpy(buf, _inv.data(), n);
+        _inv = _inv.substr(n);
+        return n;
+    }
+    virtual ssize_t recv(const struct iovec *iov, int iovcnt, int flags = 0) override {
+        while(!iov->iov_base || !iov->iov_len)
+            if (iovcnt) { ++iov, --iovcnt; }
+            else return -1;
+        return recv(iov->iov_base, iov->iov_len);
+    }
+    virtual ssize_t send(const void *buf, size_t count, int flags = 0) override {
+        size_t n = rand() % 8 + 1;
+        if (n > count) n = count;
+        auto p = (const char*)buf;
+        _out.append(p, p + n);
+        return n;
+    }
+    virtual ssize_t send(const struct iovec *iov, int iovcnt, int flags = 0) override {
+        while(!iov->iov_base || !iov->iov_len)
+            if (iovcnt) { ++iov, --iovcnt; }
+            else return -1;
+        return send(iov->iov_base, iov->iov_len);
+    }
+    virtual ssize_t sendfile(int in_fd, off_t offset, size_t count) override { return 0; }
+    virtual int close() override { return 0; }
+    virtual ssize_t read(void *buf, size_t count) override {
+        if (count > _inv.size()) count = _inv.size();
+        memcpy(buf, _inv.data(), count);
+        return count;
+    }
+    virtual ssize_t readv(const struct iovec *iov, int iovcnt) override {
+        ssize_t s = 0;
+        for (int i = 0; i < iovcnt; ++i) {
+            ssize_t ret = read(iov[i].iov_base, iov[i].iov_len);
+            if (ret < 0) return ret;
+            s += ret;
+            if (ret < (ssize_t)iov[i].iov_len) break;
+        }
+        return s;
+    }
+    virtual ssize_t write(const void *buf, size_t count) override {
+        auto p = (const char*)buf;
+        _out.append(p, p + count);
+        return count;
+    }
+    virtual ssize_t writev(const struct iovec *iov, int iovcnt) override {
+        ssize_t s = 0;
+        for (int i = 0; i < iovcnt; ++i) {
+            ssize_t ret = write(iov[i].iov_base, iov[i].iov_len);
+            if (ret < 0) return ret;
```

Comment:
should return `s` if `s > 0`?

Reply:
in the case of an error, just return the error